Improve GDScript documentation generation & behavior

anvilfolk opened the pull request:
Rework of GDScript classes' documentation generation (docgen).

### Changes:
• Removes docgen code & datastructures from `gdscript_compiler.cpp` and `gdscript.cpp`
• Docgen uses analyzer datatypes instead of less expressive runtime datatypes
• Enums return/parameter types docgen'd properly instead of being `Variant` or `int`. Fixes #71117 
• Makes hyperlinks to unopened classes generate their documentation on-demand rather than show blank screen. ~~Fixes #72406~~
• Signals parameter types are docgen'd correctly instead of being `Variant`
• Added same-line docstrings for variables/constants
• Show external class specifiers e.g. `TileSet.CellNeighbor` instead of `CellNeighbor`, `OtherClass.Enum` or `InnerClass.Enum` instead of just `Enum`
• ⚠️ ⚠️ ⚠️ Remove unnecessary class specifier path, e.g. `InnerClass.Enum` instead of `CurrentClass.InnerClass.Enum` or `Inner3.Enum` instead of `Class.Inner1.Inner2.Inner3.Enum`, when documented inside `Inner2`⚠️ ⚠️ ⚠️  this one might be slightly contentious ⚠️ ⚠️ ⚠️ 
• Created a fairly comprehensive set of gdscript files with all/most combinations of types to make sure this works properly (linked below)

### Minor changes:
• Removes unused `p_scroll` parameter in `EditorHelp::go_to_class()` and `p_vscr` in `EditorHelp::_goto_desc()`
• Removed unused `DocData::EnumDoc`
• Normalize docstrings in `EditorHelp` source code


### Enums and class specifiers everywhere!

Before (boooooo): 
![image](https://user-images.githubusercontent.com/1133892/229649853-0d9c21d6-27e9-43e9-9cee-1ede85a5b056.png)

After (yaaaaaay):
![image](https://user-images.githubusercontent.com/1133892/229649885-f1801abe-172f-42c5-b944-390fda7cbf43.png)


### Inline docstrings!

![image](https://user-images.githubusercontent.com/1133892/227028178-473d2a0f-a9c6-4c5c-80d3-81e3e4a154ca.png)

### Possibly contentious class specifier prefix elimination!

![image](https://user-images.githubusercontent.com/1133892/229648896-3e7487fd-37b5-4a71-8392-491a9ef8b69c.png)

(see below for gdscript class & enum types)

Notice how we are within `Doc1.IC`, and therefore in the highlighted areas only relevant class specifier prefixes are shown. We don't need `Doc1.IC`, just `IC`. We don't need `Doc1.IC.IE` or `IC.IE`, just `IE`. We don't need `Doc1.IC.IIC` or `Doc1.IC.IIC.IIE`, just `IIC` or `IIC.IIE`.

### Test files!

[gdscript_docgen.zip](https://github.com/godotengine/godot/files/11143025/gdscript_docgen.zip)

These are the files I used to test. They are organized as follows:

1. Two outer classes, `Doc1` and `Doc2`
2. `Doc1` and `Doc2` have an inner class `IC`, which has an inner inner class `IIC`.
3. Each class, outer or inner, has corresponding enums, `E`, `IE`, and `IIE`.
4. `Doc1` and inner classes, have methods containing all combinations of accessing classes and enums I could think of, both as parameters and return types.
 
As far as I can tell, every type name is as I thought it made the most sense (i.e. remove prefixes which are "obvious"), and click-linking works also.



<!--


Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

anvilfolk commented:
Also, for the folks following, I updated the original post with all the things happening in this PR :)

Let me know if there are any other wants that I could include here!

anvilfolk commented on editor/editor_help.cpp:
```diff
@@ -43,8 +44,53 @@
+		return false;
+	}
+
+	Ref<Script> script = ResourceLoader::load(ScriptServer::get_global_class_path(outer_class), outer_class);
+	// Check whether docgen happens in ResourceLoader::load() / whether docs are already cached
+	// If not, call reload() which docgens for GDScript
+	// TODO: could be improved with a docgen-only method, since reload() also compiles GDScripts
+	if (script.is_valid() && (script->are_docs_generated() || (script->reload(true) == OK && script->are_docs_generated()))) {
+		Vector<DocData::ClassDoc> docs = script->get_documentation();
```
So upon reflection and digging in a little more, it seems that `ResourceLoader::load()` will load the full script, which _for GDScript_ already calls `reload()` and generates documentation. So I believe we might be able to loosen and simplify these checks!

YuriSizov replied:
Yes, unless the script is already cached, it should call `reload`, which calls the compiler, which should generate the docs. So at the very least, the new `are_docs_generated` method and the attempt at reloading it here seem unnecessary?

anvilfolk replied:
Yes, it does! I have removed the changes to `Script` and its descendants, which also makes compiling this PR much faster! :)

For posterity, some more thoughts:

The alternative to `ResourceLoader` is to use the stuff in `gdscript_cache.h`, but that is not script-agnostic. However, other script languages don't have similar loading capabilities AFAIK, and because this "shallow" loading cannot happen in the `Script` class and its derivatives, it'd be hard/impossible to abstract it into that.

In the end I think the best compromise is to take the slight inefficiency of using the `ResourceLoader` abstraction, which unnecessarily compiles gdscripts in order to get their documentation.

YuriSizov commented on editor/editor_help.cpp:
```diff
@@ -43,8 +44,53 @@
 #define CONTRIBUTE_URL vformat("%s/contributing/documentation/updating_the_class_reference.html", VERSION_DOCS_URL)
 
+// TODO: this is sometimes used directly as doc->something, other times as EditorHelp::get_doc_data(); Might this be a problem?
 DocTools *EditorHelp::doc = nullptr;
```
I guess this is done on purpose? Using the method ensures that the thread used for help generation has finished, but accessing it directly, obviously, doesn't. So I think that if the help generation does it, then it's okay, but if something else does it, it's probably not okay.

For explicitness' sake I would avoid accessing it directly and add a `_get_doc_data_unsafe()` method, or something like that, for places that need access without checking for the thread,

YuriSizov replied:
But this can probably done in a follow-up, as if it's a problem, it's pre-existing.

anvilfolk replied:
Yeah, I've not really touched on any threading stuff with Godot so I'm not 100% certain whether there is even a problem in the first place. Do you think it's appropriate to leave the comment there?

YuriSizov replied:
Yes, it's good to have a comment here pointing at a potential issue.

YuriSizov commented on editor/editor_help.cpp:
```diff
@@ -225,35 +272,54 @@ void EditorHelp::_add_type(const String &p_type, const String &p_enum) {
+			display_t = link_t;
+		} else { // There is a class specifier
+			// Replace :: from enum's use of fully qualified class names with regular .
+			slices.set(0, slices[0].replace("::", "."));
```
Where is it coming from formatted like this? Can we ensure this doesn't happen?

And is it only affecting the first class in complex identifiers? Perhaps we can just run a replace on the original string?

anvilfolk replied:
Wonderful comment, this allowed me to simplify the code a whole lot! It's way simpler now!!! Thank you!

YuriSizov commented on editor/editor_help.cpp:
```diff
@@ -225,35 +272,54 @@ void EditorHelp::_add_type(const String &p_type, const String &p_enum) {
+	} else {
+		// Remove current class specifier from original type description
+		display_t = _contextualize_class_specifier(p_type, edited_class, true);
+		if (display_t.is_empty()) {
```
I don't think it's possible for this to be empty at this point, with keep last being set to true, is it? I'm not against an extra check regardless, but perhaps would add a comment that this is unlikely and is done just in case.

anvilfolk replied:
You are right! :) It was a last minute change to the contextualize function and I forgot to remove those lines! Nice catch!

YuriSizov commented on editor/editor_help.cpp:
```diff
@@ -43,8 +44,53 @@
+
+	return false;
+}
+// Removes unecessary prefix from p_class_specifier when within the p_edited_class context
```
Spellcheck reports a typo in "unnecessary" :) Also missing an empty line between the methods.

YuriSizov commented on editor/editor_help.cpp:
```diff
@@ -460,8 +526,9 @@ void EditorHelp::_pop_code_font() {
-Error EditorHelp::_goto_desc(const String &p_class, int p_vscr) {
-	if (!doc->class_list.has(p_class)) {
+Error EditorHelp::_goto_desc(const String &p_class) {
+	// If class doesn't have docs listed, attempt on-demand docgen
+	if (!doc->class_list.has(p_class) && !_attempt_doc_load(p_class)) {
```
So this is an example of `doc` being accessed without checking for the thread, which seems to be problematic already. But now we're also trying to load a script, parse the docs, and then generate editor help. So this is either going to hitch, or going to fail while threads finish.

Perhaps this is not critical at the moment, as it shouldn't be _worse_ than now, but further testing may be needed.

anvilfolk replied:
For right now I'm leaving the comments so someone with better threading knowledge than me can take a look in the future and marking this as resolved for the purposes of this PR :)

YuriSizov commented on modules/gdscript/editor/gdscript_docgen.h:
```diff
@@ -0,0 +1,42 @@
+#include "../gdscript_parser.h"
+#include "core/doc_data.h"
+
+class GDScriptDocGen {
+public:
+	static void generate_docs(GDScript *p_script, const GDScriptParser::ClassNode *p_class, bool p_inner = false);
+};
```
Does it need to be a class? Can be just a standalone method, we do this sometimes.

Edit: Though I can see that you made it a friend of other GDScript classes, so perhaps it needs scope access? Then I guess it's okay as a class.

anvilfolk replied:
I was trying to remember why I did this and I think I remember :) I'm sure it's a common C++ problem, but this was the only solution I could find but maybe there are better ones!

Here's a rundown of why I think this is necessary :)

1. This method needs access to the `ClassNode` in order to use the analyzer's datatypes. This means that wherever this method is declared needs to include `gdscript_parser.h`. Same problem if we store the `ClassNode` as a member of `GDScript` to avoid having it in the method signature.
2. If we make this a `GDScript` class method, then in order for the method to have this signature, the `gdscript.h` file needs to include `gdscript_parser.h`, which includes `gdscript_cache.h`, which includes `gdscript.h`, which is a circular include loop.
3. If we make this into a static method inside the `gdscript.cpp` file, we no longer have an include loop, but now we also don't have access to the `GDScript` class' documentation-related internals, so most of the `p_script->...` accesses would no longer work.
4. We inline the entire docgen code in `GDScript::reload()` 🤐 

Is there another way?

YuriSizov replied:
I'm not sure if there is another, more preferred way, but I like your justification :)

YuriSizov commented on modules/gdscript/gdscript.cpp:
```diff
@@ -340,12 +341,11 @@ void GDScript::_get_script_property_list(List<PropertyInfo> *r_list, bool p_incl
 			break;
 		}
 
+		props.clear();
```
Moving this is a superficial change, right? Doesn't seem like it would affect anything.

anvilfolk replied:
Correct, it's just more in line with other methods :)

YuriSizov commented on modules/gdscript/gdscript.cpp:
```diff
@@ -905,6 +745,15 @@ Error GDScript::reload(bool p_keep_state) {
+#ifdef TOOLS_ENABLED
+	// Done after compilation because it needs the GDScript object's inner class GDScript objects,
+	// which are made by calling make_scripts() within compiler.compile() above.
+	if (parser.get_tree()->identifier) { // Only generate documentation for named classes.
```
Why, though? Currently we generate it for unnamed classes as well, IIRC. Does it not work well? Docs for unnamed classes should be helpful for inspector tooltips at the very least.

anvilfolk replied:
I was really convinced this wasn't the case, but I believe you're right from a quick look through the old code. I'll dig into it a lil deeper and make sure this works as expected!

anvilfolk replied:
I can't find the code that I thought made it so we didn't generate docs for unnammed classes, which probably just means it wasn't there to begin with ;)

I removed this check :)

YuriSizov commented on modules/gdscript/gdscript.h:
```diff
@@ -83,6 +83,7 @@ class GDScript : public Script {
 	friend class GDScriptFunction;
 	friend class GDScriptAnalyzer;
 	friend class GDScriptCompiler;
+	friend class GDScriptDocGen;
```
As mentioned above, if this needs to be a class for some scope access, then this is fine. But if it doesn't, then this is not needed either.

anvilfolk replied:
See above :)

YuriSizov reviewed (changes requested):
This looks great! I'll leave the bulk of GDScript changes to @godotengine/gdscript, but it seems fine to me overall. I have left a few comments here and there which need to be resolved.

I'm also not convinced we need to support same-line doc comments, but it seems cheap to maintain, so I guess it's fine.

anvilfolk commented:
Quick note regarding inline doc comments: it was bothering me when testing because it was doubling the vertical space required by variables if you decided to document it. I really do think it's worth for classes that have more than a handful of variables. And I am always available to fix any regressions :)

YuriSizov reviewed (approved):
The editor side of things and the overall approach make sense to me. Needs a review from @godotengine/gdscript still.

adamscott commented on modules/gdscript/editor/gdscript_docgen.cpp:
```diff
@@ -0,0 +1,271 @@
+
+static String _get_class_name(const GDP::ClassNode &p_class) {
+	const GDP::ClassNode *curr_class = &p_class;
+	if (!curr_class->identifier) { // All inner classes have a identifier, so this is the outer class
```
If the outer class has `class_name` set, does it have an identifier?

anvilfolk replied:
It does! :)

The logic here is that if the outer class has an identifier, that can be used because the class gets registered as a global class and that's how you access it. If the outer class doesn't have a class name, we just use the file path, since that's the only way you can uniquely identify this file!

adamscott commented on modules/gdscript/editor/gdscript_docgen.cpp:
```diff
@@ -0,0 +1,271 @@
+	DocData::ClassDoc &doc = p_script->doc;
+
+	doc.script_path = _get_script_path(p_script->get_script_path());
+	if (!p_script->name.is_empty()) {
```
Here, for the sake of readability, I would test only if it's empty, then apply `doc.name = doc.script_path`, else `doc.name = p_script->name`

adamscott commented on modules/gdscript/editor/gdscript_docgen.cpp:
```diff
@@ -0,0 +1,271 @@
+
+	doc.brief_description = p_class->doc_brief_description;
+	doc.description = p_class->doc_description;
+	for (int i = 0; i < p_class->doc_tutorials.size(); i++) {
```
Is it possible to use something like `for (Pair<StringName, StringName> doc_tutorial : p_class->doc_tutorials)) {`?

You don't seem to use the `i` variable other than to access the element you for loop.

anvilfolk replied:
This was copy-pasted from the previous iteration of the docgen, so I don't know but I'll look into it, hehe!

anvilfolk commented on modules/gdscript/editor/gdscript_docgen.cpp:
```diff
@@ -0,0 +1,271 @@
+	return pi;
+}
+
+void GDScriptDocGen::generate_docs(GDScript *p_script, const GDP::ClassNode *p_class, bool p_inner) {
```
Also, apparently I'm not using `p_inner` anywhere? So this can be removed? Reminder for myself!

anvilfolk committed:
Improve and fix GDScript documentation generation & behavior

Removes documentation generation (docgen) from the GDScript compiler to
its own file. Adds support for GDScript enums and signal parameters and
quite a few other assorted fixes and improvements.

adamscott reviewed (approved):
Reviewed during a PR meeting of the GDScript team. 

Great job ocean, you did a really good job. It was a long time coming.

@vnen said that he would review the PR more in depth, but is ok with it at glance.

vnen commented on editor/editor_help.cpp:
```diff
@@ -74,6 +77,49 @@ class DocCache : public Resource {
+		return false;
+	}
+
+	// ResourceLoader is used in order to have a script-agnostic way to load scripts.
+	// This forces GDScript to compile the code, which is unnecessary for docgen, but it's a good compromise right now.
```
This makes me think that perhaps the ScriptServer should be aware of documentation and just as the script language to provide documentation data. This would allow not only GDScript and other languages to optimize the doc gen process.

Of course this is just something to discuss later, no need to worry about it for this PR. I agree that loading the resource is a good compromise for now.

vnen reviewed (approved):
Looks good to me. Thanks for working on this!

YuriSizov commented:
This is amazing! Thanks a lot!

anvilfolk commented:
Thanks everyone for your time, comments and support 😄 ❤️

atirut-w commented:
Not sure if this is relevant, but I'd like the LSP to be integrated with docgen so we can get consistent docgen behavior.
